src/display/editor/tools.js

```diff
@@ -0,0 +1,577 @@
+      isWin: platform.includes("Win"),
+      isMac: platform.includes("Mac"),
+    });
+  }
+
+  /**
+   * Serialize an event into a string in order to match a
+   * potential key for a callback.
+   * @param {KeyboardEvent} event
+   * @returns {string}
+   */
+  #serialize(event) {
+    if (event.altKey) {
+      this.buffer.push("alt");
+    }
+    if (event.ctrlKey) {
+      this.buffer.push("ctrl");
+    }
+    if (event.metaKey) {
+      this.buffer.push("meta");
+    }
+    if (event.shiftKey) {
+      this.buffer.push("shift");
+    }
+    this.buffer.push(event.key);
+    const str = this.buffer.join("+");
+    this.buffer.length = 0;
+
+    return str;
+  }
+
+  /**
+   * Execute a callback, if any, for a given keyboard event.
+   * The page is used as `this` in the callback.
+   * @param {AnnotationEditorLayer} page.
+   * @param {KeyboardEvent} event
+   * @returns
+   */
+  exec(page, event) {
+    if (!this.allKeys.has(event.key)) {
+      return;
+    }
+    const callback = this.callbacks.get(this.#serialize(event));
+    if (!callback) {
+      return;
+    }
+    callback.bind(page)();
+    event.preventDefault();
+  }
+}
+
+/**
+ * Basic clipboard to copy/paste some editors.
+ * It has to be used as a singleton.
+ */
+class ClipboardManager {
+  constructor() {
+    this.element = null;
+  }
+
+  /**
+   * Copy an element.
+   * @param {AnnotationEditor} element
+   */
+  copy(element) {
+    this.element = element.copy();
+  }
+
+  /**
+   * Create a new element.
+   * @returns {AnnotationEditor|null}
+   */
+  paste() {
+    return this.element?.copy() || null;
+  }
+}
+
+/**
+ * A pdf has several pages and each of them when it will rendered
+ * will have an AnnotationEditorLayer which will contain the some
+ * new Annotations associated to an editor in order to modify them.
+ *
+ * This class is used to manage all the different layers, editors and
+ * some action like copy/paste, undo/redo, ...
+ */
+class AnnotationEditorUIManager {
+  #activeEditor = null;
+
+  #allEditors = new Map();
+
+  #allLayers = new Set();
+
+  #allowClick = true;
+
+  #clipboardManager = new ClipboardManager();
+
+  #commandManager = new CommandManager();
+
+  #idManager = new IdManager();
+
+  #isAllSelected = false;
+
+  #isEnabled = false;
+
+  #mode = AnnotationEditorType.NONE;
+
+  /**
+   * Get an id.
+   * @returns {string}
+   */
+  getId() {
+    return this.#idManager.getId();
+  }
+
+  /**
+   * Add a new layer for a page which will contains the editors.
+   * @param {AnnotationEditorLayer} layer
+   */
+  addLayer(layer) {
+    this.#allLayers.add(layer);
+    if (this.#isEnabled) {
+      layer.enable();
+    } else {
+      layer.disable();
+    }
+  }
+
+  /**
+   * Remove a layer.
+   * @param {AnnotationEditorLayer} layer
+   */
+  removeLayer(layer) {
+    this.#allLayers.delete(layer);
+  }
+
+  /**
+   * Change the edition mode (FreeText, Ink, ...)
+   * @param {number} mode
+   */
+  updateMode(mode) {
```

Comment:
Please make sure that this method handles `AnnotationEditorType.NONE`, and use that to disable editing.
Most likely, the method below is no longer necessary in that case.